server.go

```diff
@@ -325,46 +325,52 @@ func (sc *naSubmissionCache) bestSubmission(net addrmgr.NetAddressType) *naSubmi
 	return best
 }
 
-// peerState maintains state of inbound, persistent, outbound peers as well
+// peerState houses state of inbound, persistent, and outbound peers as well
 // as banned peers and outbound groups.
 type peerState struct {
+	sync.Mutex
+
+	// The following fields are protected by the embedded mutex.
 	inboundPeers    map[int32]*serverPeer
 	outboundPeers   map[int32]*serverPeer
 	persistentPeers map[int32]*serverPeer
 	banned          map[string]time.Time
 	outboundGroups  map[string]int
-	subCache        *naSubmissionCache
+
+	// subCache houses the network address submission cache and is protected
+	// by its own mutex.
+	subCache *naSubmissionCache
 }
 
-// ConnectionsWithIP returns the number of connections with the given IP.
-func (ps *peerState) ConnectionsWithIP(ip net.IP) int {
-	var total int
-	for _, p := range ps.inboundPeers {
-		if ip.Equal(p.NA().IP) {
-			total++
-		}
-	}
-	for _, p := range ps.outboundPeers {
-		if ip.Equal(p.NA().IP) {
-			total++
-		}
-	}
-	for _, p := range ps.persistentPeers {
-		if ip.Equal(p.NA().IP) {
-			total++
-		}
+// makePeerState returns a peer state instance that is used to maintain the
+// state of inbound, persistent, and outbound peers as well as banned peers and
+// outbound groups.
+func makePeerState() peerState {
+	return peerState{
+		inboundPeers:    make(map[int32]*serverPeer),
+		persistentPeers: make(map[int32]*serverPeer),
+		outboundPeers:   make(map[int32]*serverPeer),
+		banned:          make(map[string]time.Time),
+		outboundGroups:  make(map[string]int),
+		subCache: &naSubmissionCache{
+			cache: make(map[string]*naSubmission, maxCachedNaSubmissions),
+			limit: maxCachedNaSubmissions,
+		},
 	}
-	return total
 }
 
-// Count returns the count of all known peers.
-func (ps *peerState) Count() int {
+// count returns the count of all known peers.
+//
+// This function MUST be called with the embedded mutex locked (for reads).
```

Comment:
Correct.  I always make it a point to comment what type of access is required though in case things are ever refactored and also it makes it clear whether or not the state will be mutated.